### ISSUE TITLE
Difference between coverage results with source specifies full dir instead of module name

### ISSUE THREAD

**Issue by nedbat** [coveragepy/coveragepy#426]
*Originally reported by* **Alexander Todorov (Bitbucket: [atodorov](https://bitbucket.org/atodorov), GitHub: [atodorov](https://github.com/atodorov))**

----------------------------------------

Anaconda, the Fedora Linux installer supports coverage collection via the API interface. If you boot with `inst.debug` then the main file `/usr/sbin/anaconda` will initialize coverage as one of the first things it does. 

```
    cov = coverage.coverage(data_file="/mnt/sysimage/root/anaconda.coverage",
                            branch=True,
                            source=["/usr/sbin/anaconda", "pyanaconda"]
                            )
    cov.start()
```

The results are 105 files measured, which is short of 19 files which are also present in the pyanaconda module. OTOH if the source path is specified as the full path (e.g. `/usr/lib64/python3.4/site-packages/pyanaconda`) then all files inside are included in the coverage data. See:
https://github.com/rhinstaller/anaconda/pull/397

I've tried to reproduce locally with the same directory structure (but not the same code) and was not able to. I will keep trying so it's easier to reproduce and easier to debug/fix.

----------------------------------------
- Bitbucket: https://bitbucket.org/ned/coveragepy/issue/426

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

I tried to reproduce the problem with a foo module containing a single a.py file 

```
#!python

def main():
    print('Hello')

```
With the following we have 50% coverage and that is to be expected:

```
#!bash

$ cat ~/tmp/b.py 
from foo import a
import coverage
cov = coverage.coverage(source=["foo"])
cov.start()
a.main()
cov.stop()
cov.report()
$ PYTHONPATH=/tmp python ~/tmp/b.py 
Hello
Coverage.py warning: Module foo was previously imported, but not measured.
Name            Stmts   Miss  Cover
-----------------------------------
/tmp/foo/a.py       2      1    50%
```
Changing the file to import the module after starting coverage there is 100% coverage:

```
#!bash

$ cat ~/tmp/b.py 
import coverage
cov = coverage.coverage(source=["foo"])
cov.start()
from foo import a
a.main()
cov.stop()
cov.report()
$ PYTHONPATH=/tmp python ~/tmp/b.py 
Hello
Name                   Stmts   Miss  Cover
------------------------------------------
/tmp/foo/__init__.py       0      0   100%
/tmp/foo/a.py              2      0   100%
------------------------------------------
TOTAL                      2      0   100%
```
@atodorov am I missing something ? Note that this is with coverage master (i.e. > 4.2) and it could be that things evolved for the best since 4.0 ;-)

**Comment by nedbat** [comment]
*Original comment by* **Alexander Todorov (Bitbucket: [atodorov](https://bitbucket.org/atodorov), GitHub: [atodorov](https://github.com/atodorov))**

----------------------------------------

@dachary I don't think your comment is related to the originally reported issue.  If you take a look at the PR I've linked above you will see that I import all modules I want to cover after coverage has been enabled. My problem was the way in which the source path was specified. However I don't have the time to investigate and see how things are with 4.x

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

@atodorov I apologize for the confusion. Your [fix](https://github.com/rhinstaller/anaconda/pull/397/files) look right: the source argument is a list of directories. If you specify pyanaconda instead of the full path to the directory it will only work as expected if pyanaconda is a subdirectory of your current working directory. Am I making sense this time ?

**Comment by nedbat** [comment]
*Original comment by* **Alexander Todorov (Bitbucket: [atodorov](https://bitbucket.org/atodorov), GitHub: [atodorov](https://github.com/atodorov))**

----------------------------------------

Thanks for the update, now it makes sense and indeed I had a "pyanaconda" directory locally. I guess this is something else you'd like to document with a PR, isn't it ?

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

I'm not sure how to update the documentation. It states that the **--source** is a list of directories. Maybe there is a way to convey that more strongly ?

**Comment by nedbat** [comment]
*Original comment by* **Alexander Todorov (Bitbucket: [atodorov](https://bitbucket.org/atodorov), GitHub: [atodorov](https://github.com/atodorov))**

----------------------------------------

I guess I haven't read that :). Maybe just add the note from your previous comment about full path vs. relative directories if the later exist in the project ?

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

:-D I'm not a native english speaker, do you have a wording in mind to update [the documentation](https://coverage.readthedocs.io/en/coverage-4.2/source.html#execution) ? I'd be happy to create a pull request with it.

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

@nedbat re-reading [the documentation](https://coverage.readthedocs.io/en/coverage-4.2/source.html#execution) I see it reads **directories or package names** but I'm not sure in which case using a packaging name makes a difference ?

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

The --source is split into [source and source_pkgs](https://bitbucket.org/ned/coveragepy/src/e32f15eec624f6b979b5ae09af64446d7c7ce9f9/coverage/control.py?at=default&fileviewer=file-view-default#control.py-218) variables depending. It is then used to set [source_match and source_pkgs_match](https://bitbucket.org/ned/coveragepy/src/e32f15eec624f6b979b5ae09af64446d7c7ce9f9/coverage/control.py?at=default&fileviewer=file-view-default#control.py-324) to figure out if a [source should be taken into account](https://bitbucket.org/ned/coveragepy/src/e32f15eec624f6b979b5ae09af64446d7c7ce9f9/coverage/control.py?at=default&fileviewer=file-view-default#control.py-550) during the **run**. 

After the run and before saving the data to **.coverage**, the **--source** is used again to [find which files were never executed](https://bitbucket.org/ned/coveragepy/src/e32f15eec624f6b979b5ae09af64446d7c7ce9f9/coverage/control.py?at=default&fileviewer=file-view-default#control.py-836). But the [find_python_files](https://bitbucket.org/ned/coveragepy/src/e32f15eec624f6b979b5ae09af64446d7c7ce9f9/coverage/files.py?at=default&fileviewer=file-view-default#files.py-368) does not know how to iterate over the files of a package. It only knows about listing the files in a directory.

This difference between directories and packages in the --source argument should be documented. Or maybe this should be considered a bug and there should be no difference between files and modules.

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

@atodorov Here is a tentative fix so there is no difference between directories and modules https://bitbucket.org/ned/coveragepy/pull-requests/122/make-source-module-do-the-same-as-source/diff

The alternative would be to document the difference.

What do you think ?

**Comment by nedbat** [comment]
Do I understand this correctly? This bug is about a difference between source=dir and source=pkg, and the difference is that with source=dir, coverage will find completely unexecuted files, but with source=pkg, it will not.

Right?

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

Right.

**Comment by nedbat** [comment]
make --source module do the same as --source directory #426

The --source argument can either be a module or a directory. The user
expects that it behaves the same in both cases. Make sure the module
is recursively explored so that files that are not run show in the
coverage report.

close #426

→ <<cset [ed784d8f334e (bb)](https://bitbucket.org/ned/coveragepy/commits/ed784d8f334e)>>

**Comment by nedbat** [comment]
This was shipped in 4.3.2.
